benchmarks/dataset_serialize_benchmark.py

```diff
@@ -0,0 +1,232 @@
+import itertools
+import logging
+import os
+import shutil
+import time
+import uuid
+
+import conbench.runner
+import pyarrow
+import pyarrow.dataset as ds
+
+from benchmarks import _benchmark
+
+log = logging.getLogger(__name__)
+
+
+@conbench.runner.register_benchmark
+class DatasetSerializeBenchmark(_benchmark.Benchmark):
+    """
+    This benchmark is supposed to measure the time it takes to write data from
+    memory (from a pyarrow Table) to a tmpfs file system, given a specific
+    serialization format (parquet, arrow, ...).
+
+    To make this benchmark agnostic to disk read performance on the input side
+    of things, the data is read fully into memory before starting (and timing)
+    the benchmark function. That is (believed to be) achieved with:
+
+        data = source_dataset.to_table(filter=somefilter, ...)
+
+    After data of interest has been read into memory, the following call is
+    used for both serialization and writing-to-filesystem in one go:
+
+        pyarrow.dataset.write_dataset(format=someformat, ...)
+
+    That operation is timed (and the duration is the major output of this
+    benchmark).
+
+    The data is written to `/dev/shm` (available on all Linux systems). This is
+    a file system backed by RAM (tmpfs). The assumption is that writing to
+    tmpfs is fast (so fast that benchmark duration is significantly affected by
+    the serialization itself), and stable (so that its performance is ~constant
+    across runs on the same machine).
+
+    This benchmark does not resolve how much time goes into the CPU work for
+    serialization vs. the system calls for writing to tmpfs (that would be a
+    different question to answer, an interesting one, that is maybe more of a
+    task for profiling).
+
+    There are two dimensions that are varied:
+
+        - serialization format
+        - amount of the data being written, as set by a filter on the input
+    """
+
+    name = "dataset-serialize"
+
+    arguments = ["source"]
+
+    sources = [
+        "nyctaxi_multi_parquet_s3",
+        # "nyctaxi_multi_ipc_s3",
+        # "chi_traffic_2020_Q1",
+    ]
+
+    sources_test = [
+        "nyctaxi_multi_parquet_s3_sample",
+        "nyctaxi_multi_ipc_s3_sample",
+        "chi_traffic_sample",
+    ]
+
+    _params = {
+        "selectivity": ("1pc", "10pc"),
```

Comment:
In the end we'd also like to measure 100%, is that not included to keep the runtime in check?

Reply:
I am conservative here to keep /dev/shm usage in check.

On my machine I have a 16 GB /dev/shm and I managed to fill this and almost crashed my desktop environment. 

I will play a little more with that and maybe update the PR again to add a variant that writes more data, maybe not 100 %, but more than 10 % :).

Reply:
Added commits. For `nyctaxi_multi_ipc_s3` and the `10 %` / `csv` scenario the output is already ~ 5 GB:

```
[221214-17:11:21.205] [375153] [benchmarks.dataset_serialize_benchmark] INFO: case ('10pc', 'csv'): create directory
[221214-17:11:21.205] [375153] [benchmarks.dataset_serialize_benchmark] INFO: directory created, path: /dev/shm/bench-02a3a0e3/10pc-csv-de09c188-69d7-41b4-9c0b-fd00654f06ad
[221214-17:11:21.877] [375153] [benchmarks.dataset_serialize_benchmark] INFO: read source dataset into memory in 0.6720 s
[221214-17:11:50.099] [375153] [benchmarks.dataset_serialize_benchmark] INFO: stdout of ['du', '-sh', '/dev/shm/bench-02a3a0e3/10pc-csv-de09c188-69d7-41b4-9c0b-fd00654f06ad']: 4.6G
```

I think we should not grow beyond that

Reply:
Update: I have adopted the `‘overwrite_or_ignore’` technique and enabled the `100 %` case. The maximum usage of /dev/shm now is 7.6 GB, and I think that's OK. (if the CI runners also provide that much of space).